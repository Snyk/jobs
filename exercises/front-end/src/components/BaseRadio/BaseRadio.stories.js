import '~/assets/scss/styles.scss';
import BaseRadio from './BaseRadio';

export default {
  title: 'Library/Base Components/Radio',
  component: BaseRadio,
};

export const Default = () => ({
  components: { BaseRadio },
  template: `
    <div>
      <BaseRadio
        label="Default"
        name="group-1"
        value="1"
      />

      <BaseRadio
        label="Checked"
        name="group-2"
        value="1"
        checked
      />

      <BaseRadio
        label="Disabled"
        name="group-3"
        value="1"
        disabled
      />

      <BaseRadio
        label="Disabled and Checked"
        name="group-4"
        value="1"
        disabled
        checked
      />
    </div>
  `,
});

export const WithHTML = () => ({
  components: { BaseRadio },
  template: `
    <BaseRadio name="group-5" value="1" checked>
      BaseBadge with <a href="#">html inside</a>
    </BaseRadio>
  `,
});
WithHTML.storyName = 'â€¦with HTML';
