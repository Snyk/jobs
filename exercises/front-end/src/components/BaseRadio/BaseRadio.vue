<template>
  <LayoutFormElement :class="computedClasses" class="vue--radio" spacing="none">
    <label>
      {{ label }}
      <slot />
      <input
        :id="id"
        :name="name"
        :value="content"
        :disabled="disabled"
        :checked="isChecked"
        type="radio"
        class="vue--radio__input"
        @change="onChange"
      />
      <span
        class="vue--radio__toggle"
        tabindex="0"
        @keyup.enter="onChange"
        @keyup.space="onChange"
      />
    </label>
  </LayoutFormElement>
</template>

<script>
import LayoutFormElement from '~/components/LayoutFormElement/LayoutFormElement';

export default {
  name: 'BaseRadio',
  status: 'under-review',

  components: { LayoutFormElement },

  props: {
    /**
     * ID of the checkbox input tag
     */
    id: {
      type: String,
      default: null,
    },
    /**
     * BaseBadge for the checkbox
     */
    label: {
      type: String,
      default: null,
    },
    /**
     * Name of the checkbox input tag
     */
    name: {
      type: String,
      default: null,
    },
    /**
     * Value of the checkbox input tag
     */
    value: {
      type: String,
      default: null,
    },
    /**
     * Whether the checkbox is disabled or not
     */
    disabled: {
      type: Boolean,
      default: false,
    },
    /**
     * Whether the checkbox is checked or not
     */
    checked: {
      type: Boolean,
      default: false,
    },
  },

  data: function () {
    return {
      isChecked: this.checked,
      content: this.value,
    };
  },

  computed: {
    computedClasses() {
      return {
        'vue--radio--disabled': this.disabled,
      };
    },
  },

  created() {
    this.$root.$on('group-change', (value, name) => {
      if (this.name === name && this.content !== value) {
        this.isChecked = false;
      }
    });
  },

  methods: {
    onChange() {
      this.$root.$emit('group-change', this.content, this.name);
      this.isChecked = !this.isChecked;
      this.$emit('change', this.content, this.name);
    },
  },
};
</script>

<style lang="scss" scoped>
@import '~/assets/scss/utils';

.vue--radio {
  $self: &;

  display: block;
  line-height: rem(30px);
  padding: 0 0 0 1.5em;
  position: relative;

  &__toggle {
    @include fieldStyles();
    @include hoverState();

    border-radius: 50%;
    height: 1em;
    left: 0;
    padding: 0;
    position: absolute;
    top: rem(6px);
    width: 1em;
  }

  &__input {
    display: none;

    &:checked ~ #{$self}__toggle {
      box-shadow: inset 0 0 0 1px color(action), 0 0 0 2px transparent,
        inset 0 0 0 4px color(ui, white), inset 0 0 0 8px color(action);
      transition: box-shadow 0.2s ease;

      &:focus,
      &:hover {
        box-shadow: inset 0 0 0 1px color(action), 0 0 0 1px color(action),
          inset 0 0 0 4px color(ui, white), inset 0 0 0 8px color(action);
        outline: none;
      }
    }
  }

  &--disabled {
    #{$self}__toggle {
      @include disabledState();
    }

    #{$self}__input:checked ~ #{$self}__toggle {
      border: none;
      box-shadow: inset 0 0 0 1px color(disabled, border), 0 0 0 2px transparent,
        inset 0 0 0 4px color(disabled, bg),
        inset 0 0 0 8px color(disabled, border);
    }
  }
}
</style>
