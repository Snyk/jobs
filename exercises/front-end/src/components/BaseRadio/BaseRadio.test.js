import { shallowMount } from '@vue/test-utils';
import BaseRadio from '~/components/BaseRadio/BaseRadio';

describe('BaseRadio', () => {
  describe('Snapshots', () => {
    it('renders as expected', () => {
      const wrapper = shallowMount(BaseRadio, {
        propsData: {
          name: 'radio',
          label: "I'm a radio input",
        },
      });

      expect(wrapper).toMatchSnapshot();
    });

    it('renders valid html', () => {
      const wrapper = shallowMount(BaseRadio, {
        propsData: {
          name: 'radio',
          label: "I'm a radio input",
        },
      });

      expect(wrapper.html()).toHTMLValidate();
    });
  });

  describe('Defaults', () => {
    const wrapper = shallowMount(BaseRadio, {
      propsData: {
        name: 'radio',
      },
    });

    it('defaults to empty label', () => {
      expect(wrapper.props('label')).toBeNull();
    });

    it('defaults to enabled', () => {
      expect(wrapper.props('disabled')).toBe(false);
    });

    it('defaults to unchecked', () => {
      expect(wrapper.props('checked')).toBe(false);
    });
  });

  describe('Events', () => {
    const wrapper = shallowMount(BaseRadio, {
      propsData: {
        name: 'radio',
        value: 'dummy-value',
      },
    });

    it('emits a change event', async () => {
      const radiobutton = wrapper.find('input');
      await radiobutton.trigger('change');
      expect(wrapper.emitted('change')).toHaveLength(1);
      expect(wrapper.emitted('change')[0][0]).toEqual('dummy-value');
      expect(wrapper.emitted('change')[0][1]).toEqual('radio');
    });
  });
});
