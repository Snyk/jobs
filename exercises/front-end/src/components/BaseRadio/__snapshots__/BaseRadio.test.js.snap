// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseRadio Snapshots renders as expected 1`] = `
<layoutformelement-stub spacing="none" class="vue--radio"><label>
    I'm a radio input
    <input name="radio" type="radio" class="vue--radio__input" value=""> <span tabindex="0" class="vue--radio__toggle"></span></label></layoutformelement-stub>
`;
